tango/integrations/wandb/workspace.py

```diff
@@ -0,0 +1,427 @@
+from tango.workspace import Run, Workspace
+
+from .step_cache import WandbStepCache
+from .util import ArtifactKind, RunKind, check_environment
+
+T = TypeVar("T")
+
+logger = logging.getLogger(__name__)
+
+
+@Workspace.register("wandb")
+class WandbWorkspace(Workspace):
+    """
+    This is a :class:`~tango.workspace.Workspace` that tracks Tango runs in a W&B project.
+    It also stores step results as W&B Artifacts via :class:`WandbStepCache`.
+
+    Each Tango run with this workspace will generate multiple runs in your W&B project.
+    There will always be a W&B run corresponding to each Tango run with the same name,
+    which will contain some metadata about the Tango run. Then there will be one W&B run
+    for each cacheable step that runs with a name corresponding to the name of the step.
+    So if your Tango run includes 3 cacheable steps, that will result in a total of 4 new runs in W&B.
+
+    :param project: The W&B project to use for the workspace.
+    :param entity: The W&B entity (user or organization account) to use for the workspace.
+
+    .. tip::
+        Registered as a :class:`~tango.workspace.Workspace` under the name "wandb".
+    """
+
+    def __init__(self, project: str, entity: Optional[str] = None):
+        check_environment()
+        super().__init__()
+        self.project = project
+        self._entity = entity
+        self.cache = WandbStepCache(project=self.project, entity=self.entity)
+        self.steps_dir = tango_cache_dir() / "wandb_workspace"
+        self.locks: Dict[Step, FileLock] = {}
+        self._running_step_info: Dict[str, StepInfo] = {}
+
+    def __getstate__(self):
+        """
+        We override `__getstate__()` to customize how instances of this class are pickled
+        since we don't want to persist certain attributes.
+        """
+        out = super().__getstate__()
+        out["locks"] = {}
+        return out
+
+    @property
+    def wandb_client(self) -> wandb.Api:
+        overrides = {"project": self.project}
+        if self._entity is not None:
+            overrides["entity"] = self._entity
+        return wandb.Api(overrides=overrides)
+
+    @property
+    def entity(self) -> str:
+        return self._entity or self.wandb_client.default_entity
+
+    @property
+    def url(self) -> str:
+        return f"wandb://{self.entity}/{self.project}"
+
+    @classmethod
+    def from_parsed_url(cls, parsed_url: ParseResult) -> Workspace:
+        entity = parsed_url.netloc
+        project = parsed_url.path
+        if project:
+            project = project.strip("/")
+        return cls(project=project, entity=entity)
+
+    @property
+    def step_cache(self) -> StepCache:
+        return self.cache
+
+    @property
+    def wandb_project_url(self) -> str:
+        """
+        The URL of the W&B project this workspace uses.
+        """
+        app_url = self.wandb_client.client.app_url
+        if app_url.endswith("/"):
+            app_url = app_url[:-1]
```

Comment:
```suggestion
        app_url = app_url.rstrip("/")
```
This will also handle multiple trailing slashes.